daemon/actual_ip_resolver.go

```diff
@@ -0,0 +1,131 @@
+package daemon
+
+import (
+	"context"
+	"log"
+	"net/netip"
+	"time"
+
+	"github.com/NordSecurity/nordvpn-linux/core"
+	"github.com/NordSecurity/nordvpn-linux/daemon/state"
+	"github.com/NordSecurity/nordvpn-linux/events"
+	"github.com/NordSecurity/nordvpn-linux/internal"
+	"github.com/NordSecurity/nordvpn-linux/network"
+)
+
+// tryInsightsIP is an attempt to get insights IP
+// the IP it returns is always valid
+func tryInsightsIP(api core.InsightsAPI) (netip.Addr, error) {
+	insights, err := api.InsightsViaTunnel()
+	if err != nil || insights == nil {
+		return netip.Addr{}, err
+	}
+
+	ip, err := netip.ParseAddr(insights.IP)
+	if err == nil {
+		return ip, nil
+	} else {
+		return netip.Addr{}, err
+	}
+}
+
+func insightsIPUntilSuccess(ctx context.Context, api core.InsightsAPI, backoff func(int) time.Duration) (netip.Addr, error) {
+	type Result struct {
+		netip.Addr
+		error
+	}
+	result := make(chan Result, 1)
+	for i := 0; ; i++ {
+		if ctx.Err() != nil {
+			return netip.Addr{}, ctx.Err()
+		}
+
+		// this goroutine is used so that this function is immediately stopped once the context is cancelled
+		go func() {
+			ip, err := tryInsightsIP(api)
+			if ctx.Err() == nil {
```

Comment:
Won't this cause a race? :thinking: if the context gets cancelled after this `if` but before sending data via `result` channel, it's possible that the loop will exit, but you'll still send something to `result` and there is no one left to read it I think. That's probably not an issue though, but I think that's additional reason for not making this check in a separate routine

Reply:
i don't see any way to make it to not be in a separate goroutine.
It has to stop when the context is cancelled.
if you send to channel, but no one is listening, simply nothing will happen.